File: haskell/grade-school/School.hs
module School(School, empty, add, sorted, grade) where

import Data.List(sort)

data School = Empty | Grade Int [String] School
type Grade = (Int, [String])

empty :: School
empty = Empty

add :: Int -> String -> School -> School
add gradeNum student Empty = Grade gradeNum [student] empty
add gradeNum student (Grade thisGrade list school)
    | gradeNum == thisGrade = Grade thisGrade (student:list) school
    | otherwise = Grade thisGrade list (add gradeNum student school)

sorted :: School -> [Grade]
sorted Empty = []
sorted (Grade gradeNum list school) =
    minorGrades ++ [(gradeNum, sort list)] ++ majorGrades
    where
        sortedSchool = sorted school
        minorGrades  = [grd | grd <- sortedSchool, fst grd < gradeNum]
        majorGrades  = [grd | grd <- sortedSchool, fst grd >= gradeNum]

grade :: Int -> School -> [String]
grade _ Empty = []
grade gradeNum (Grade thisGrade list school)
    | gradeNum == thisGrade = list
    | otherwise = grade gradeNum school

File: python/grade-school/school.py
class School(object):
    def __init__(self, name=""):
        self.db = {}
        self.name = name

    def add(self, student, grade):
        if grade not in self.db:
            self.db[grade] = {student}
        else:
            self.db[grade].update({student})

    def grade(self, grade):
        try:
            return self.db[grade]
        except:
            return set()

    def sort(self):
        school = []
        for grade, students in self.db.items():
            try:
                students = list(students)
                students.sort()
                school.append((grade, tuple(students)))
            except:
                print(students)

        return school
